includes/TaxCalculation/class-order-calculation-logger.php

```diff
@@ -0,0 +1,229 @@
+			$message = $this->format_failed_calculation_message( $details );
+		} else {
+			$message = $this->format_unexpected_exception_message( $details );
+		}
+
+		$this->log( $level, $message );
+	}
+
+	/**
+	 * Determines log level of failure event.
+	 *
+	 * @param Exception $exception Exception thrown to cause failure.
+	 *
+	 * @return string Log level.
+	 */
+	private function get_failed_log_level( $exception ) {
+		if ( $this->is_taxjar_exception( $exception ) ) {
+			return WC_Log_Levels::NOTICE;
+		} else {
+			return WC_Log_Levels::ERROR;
+		}
+	}
+
+	/**
+	 * Determine if exception is a Tax_Calculation_Exception
+	 *
+	 * @param Exception $exception Exception thrown.
+	 *
+	 * @return bool
+	 */
+	private function is_taxjar_exception( $exception ) {
+		return $exception instanceof Tax_Calculation_Exception;
+	}
+
+	/**
+	 * Formats message for calculation failure.
+	 *
+	 * @param array $details Tax calculation details.
+	 *
+	 * @return string
+	 */
+	private function format_failed_calculation_message( $details ) {
+		$message  = 'TaxJar could not calculate tax on order #' . $this->order->get_id() . '. ';
+		$message .= 'Reverting to default WooCommerce tax calculation.';
+		$message .= $this->format_reason( $details['exception'] );
+		$message .= $this->format_context( $details['context'] );
+		$message .= $this->format_request_details( $details );
+		$message .= $this->format_response_details( $details );
+		$message .= PHP_EOL;
+		return $message;
+	}
+
+	/**
+	 * Formats failure reason.
+	 *
+	 * @param Exception $exception Exception thrown.
+	 *
+	 * @return string
+	 */
+	private function format_reason( $exception ) {
+		return PHP_EOL . 'Reason: ' . $exception->getMessage();
+	}
+
+	/**
+	 * Formats context of tax calculation.
+	 *
+	 * @param string $context Context of calculation.
+	 *
+	 * @return string
+	 */
+	private function format_context( $context ) {
+		return PHP_EOL . 'Context: ' . $context;
+	}
+
+	/**
+	 * Formats the request body details of tax calculation.
+	 *
+	 * @param array $details Tax calculation details.
+	 *
+	 * @return string
+	 */
+	private function format_request_details( $details ) {
+		if ( ! empty( $details['request_body'] ) ) {
+			return PHP_EOL . 'Request: ' . $details['request_body']->to_json();
+		} else {
+			return '';
+		}
+	}
+
+	/**
+	 * Formats response details from TaxJar API.
+	 *
+	 * @param array $details Tax calculation details.
+	 *
+	 * @return string
+	 */
+	private function format_response_details( $details ) {
+		if ( ! empty( $details['tax_details'] ) ) {
+			$response = $details['tax_details']->get_raw_response();
+			$message  = '';
+			$message .= $this->format_response_message( $response );
+			$message .= $this->format_response_body( $response );
+			return $message;
+		} else {
+			return '';
+		}
+	}
+
+	/**
+	 * Format response message from TaxJar API.
+	 *
+	 * @param array $response Response from TaxJar API.
+	 *
+	 * @return string
```

Comment:
Will `format_response_message` and `format_response_body` return strings if the response is empty?

Reply:
So when concatenating strings, PHP treats the value of a method with no return statement as an empty string. So these methods do work as intended. However I think we should be more consistent and use the same approach across all the related methods. I'll make that fix. Thanks!